src/vm/crossloaderallocatorhash.inl

```diff
@@ -0,0 +1,1216 @@
+    return usedEntries;
+}
+
+#ifndef DACCESS_COMPILE
+template <class TKey_, class TValue_>
+/*static*/ void NoRemoveDefaultCrossLoaderAllocatorHashTraits<TKey_, TValue_>::SetUsedEntries(TValue* pStartOfValuesData, DWORD entriesInArrayTotal, DWORD usedEntries)
+{
+    if (usedEntries < entriesInArrayTotal)
+    {
+        if (usedEntries == (entriesInArrayTotal - 1))
+        {
+            pStartOfValuesData[entriesInArrayTotal - 1] = (TValue)0;
+        }
+        else
+        {
+            pStartOfValuesData[entriesInArrayTotal - 1] = (TValue)(usedEntries);
+            pStartOfValuesData[entriesInArrayTotal - 2] = (TValue)0;
+        }
+    }
+}
+
+template <class TKey_, class TValue_>
+/*static*/ bool NoRemoveDefaultCrossLoaderAllocatorHashTraits<TKey_, TValue_>::AddToValuesInHeapMemory(OBJECTREF *pKeyValueStore, const TKey& key, const TValue& value)
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_COOPERATIVE;
+    }
+    CONTRACTL_END;
+
+    static_assert(sizeof(TKey)==sizeof(TValue), "Assume keys and values are the same size");
+
+    bool updatedKeyValueStore = false;
+
+    if (*pKeyValueStore == NULL)
+    {
+        *pKeyValueStore = AllocatePrimitiveArray(ELEMENT_TYPE_I1, (value == NULL) ? sizeof(TKey) : sizeof(TKey) + sizeof(TValue), FALSE);
+        updatedKeyValueStore = true;
+        TKey* pKeyLoc = (TKey*)((I1ARRAYREF)*pKeyValueStore)->GetDirectPointerToNonObjectElements();
+        *pKeyLoc = key;
+        if (value != NULL)
+        {
+            TValue* pValueLoc = (TValue*)(((I1ARRAYREF)*pKeyValueStore)->GetDirectPointerToNonObjectElements() + sizeof(TKey));
+            *pValueLoc = value;
+        }
+    }
+    else if (value != NULL)
+    {
+        DWORD entriesInArrayTotal;
+        DWORD usedEntries = ComputeUsedEntries(pKeyValueStore, &entriesInArrayTotal);
+
+        if (usedEntries == entriesInArrayTotal)
+        {
+            // There isn't free space. Build a new, bigger array with the existing data 
+            DWORD newSize;
+            if (usedEntries < 8)
+                newSize = usedEntries + 1; // Grow very slowly initially. The cost of allocation/copy is cheap, and this holds very tight on memory usage
+            else
+                newSize = usedEntries * 2;
+
+            if (newSize < usedEntries)
+                COMPlusThrow(kOverflowException);
+
+            // Allocate the new array.
+            I1ARRAYREF newKeyValueStore = (I1ARRAYREF)AllocatePrimitiveArray(ELEMENT_TYPE_I1, newSize*sizeof(TValue) + sizeof(TKey), FALSE);
+
+            // Since, AllocatePrimitiveArray may have triggered a GC, recapture all data pointers from GC objects
+            void* pStartOfNewArray = newKeyValueStore->GetDirectPointerToNonObjectElements();
+            void* pStartOfOldArray = ((I1ARRAYREF)*pKeyValueStore)->GetDirectPointerToNonObjectElements();
+
+            memcpyNoGCRefs(pStartOfNewArray, pStartOfOldArray, ((I1ARRAYREF)*pKeyValueStore)->GetNumComponents());
```

Comment:
Looks like memcpyNoGCRefs just calls memcpy with the same arguments. Should the last argument be size in bytes? Does `GetNumComponents()` get the number of elements in the array or size in bytes?

Also it looks like this and other tests for `NULL` on the value imply that the key and value must be pointer-sized types. Maybe add a `static_assert` if there isn't one already, next to the one above.

Reply:
GetNumComponents for an I1Array (array of sbytes) returns the number of elements in the array, which is the same as the number of bytes.

About the NULL checks, the intention is to be able to hold pointer like values that are comparable to NULL, but not necessarily require the data to be stored as a pointer. David Mason and I have discussed making this data structure particularly efficient for his purposes (and there are additional works necessary to pull of that efficiency, but one of them is to use 48bit pointers instead of 64bit pointers. (It turns out all of our platforms that we expect to support in the short or medium term restrict pointer support to only actually supporting a 48 bit address space. I'm not going to write in that support now, but if it turns out its useful, I'd like to leave us free to have 48bit pointers.

Reply:
I've added the IsNull method as suggested in another comment.